From patchwork Tue Sep  6 19:49:02 2022
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Liam Howlett <Liam.Howlett@Oracle.com>
X-Patchwork-Id: 12968096
Return-Path: <owner-linux-mm@kvack.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from kanga.kvack.org (kanga.kvack.org [205.233.56.17])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 92E37ECAAA1
	for <linux-mm@archiver.kernel.org>; Tue,  6 Sep 2022 19:50:11 +0000 (UTC)
Received: by kanga.kvack.org (Postfix)
	id 5E4798D000F; Tue,  6 Sep 2022 15:49:54 -0400 (EDT)
Received: by kanga.kvack.org (Postfix, from userid 40)
	id 4F65E8D000A; Tue,  6 Sep 2022 15:49:54 -0400 (EDT)
X-Delivered-To: int-list-linux-mm@kvack.org
Received: by kanga.kvack.org (Postfix, from userid 63042)
	id 1EB398D000E; Tue,  6 Sep 2022 15:49:54 -0400 (EDT)
X-Delivered-To: linux-mm@kvack.org
Received: from relay.hostedemail.com (smtprelay0014.hostedemail.com
 [216.40.44.14])
	by kanga.kvack.org (Postfix) with ESMTP id E27548D000A
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 15:49:53 -0400 (EDT)
Received: from smtpin06.hostedemail.com (a10.router.float.18 [10.200.18.1])
	by unirelay03.hostedemail.com (Postfix) with ESMTP id C872FA091C
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:49:53 +0000 (UTC)
X-FDA: 79882701066.06.85DF5CE
Received: from mx0a-00069f02.pphosted.com (mx0a-00069f02.pphosted.com
 [205.220.165.32])
	by imf14.hostedemail.com (Postfix) with ESMTP id 70D30100069
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:49:53 +0000 (UTC)
Received: from pps.filterd (m0246617.ppops.net [127.0.0.1])
	by mx0b-00069f02.pphosted.com (8.17.1.5/8.17.1.5) with ESMTP id
 286IdYjZ029914;
	Tue, 6 Sep 2022 19:49:52 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : references : in-reply-to : content-type :
 content-transfer-encoding : mime-version; s=corp-2022-7-12;
 bh=7N8JCa2g2lf1em2rWSLx//NSbUzOTVVgGzz1o4lCCv4=;
 b=hh3v+HEPVNCDjb7Co3GGwFtVHVRvjnxMsj2i/HtXJ7W9MbpB3kDVnPakan9z3sgYsD8P
 1dqdjlpU6hvIHJHMsHZaF6QIqm0hcP/5nXduqaArkd7wtY9ZANMi1mZsJXR9T5qdfrVt
 d2yejkUPwml9ekq1pr0HekurNYqtuFRmGHWRY1v9DSSGxgQf6y65DHNcM0JKuZZwGvMB
 5CZ5BRJsjozZVbW9rUCvQZ08ptwx7Pcmd2PxlsFSYX9bghnvsjrFzUeBJ5L6v/cp0vBY
 /BbtMf2pj2IVwrrHIKuRPW5SI2jXQC1ocpKUNAtkd0kib5u2e1C4sAnt6iY1IbLkEoPM 2Q==
Received: from phxpaimrmta01.imrmtpd1.prodappphxaev1.oraclevcn.com
 (phxpaimrmta01.appoci.oracle.com [138.1.114.2])
	by mx0b-00069f02.pphosted.com (PPS) with ESMTPS id 3jbyftpvdg-1
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:49:52 +0000
Received: from pps.filterd
 (phxpaimrmta01.imrmtpd1.prodappphxaev1.oraclevcn.com [127.0.0.1])
	by phxpaimrmta01.imrmtpd1.prodappphxaev1.oraclevcn.com (8.17.1.5/8.17.1.5)
 with ESMTP id 286HF67X031251;
	Tue, 6 Sep 2022 19:49:51 GMT
Received: from nam11-co1-obe.outbound.protection.outlook.com
 (mail-co1nam11lp2173.outbound.protection.outlook.com [104.47.56.173])
	by phxpaimrmta01.imrmtpd1.prodappphxaev1.oraclevcn.com (PPS) with ESMTPS id
 3jbwc3ahkd-10
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:49:51 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=UeXtOnZ37V31r4RTwKFyqALKSMrtS0ZccUA1WQYDZnHDplvHcPkXaSsQU2cDpLZBs5T6J9NbLWWDCd449FjOMDSuaCuFKw1uIutELLPDcowRQvXFYryw8uBk8lnUdoBEg84r+UhAXS/5NJr9v31Xz/nJmMA1j3TCpE808HobwMAih7dnuW04kCEMEsIHnDqA1RBFC7YHjjlxyVdar84tgepPebe+paEqJ/XnXw2/USirfpROemr6TqYnoS0cSFoTtf/LB9tyA+r7jSWAjHt/WUiquoh0QJa7TEhtSjfQbUt5fYoDajzJHVMmAbrUypWMiHnsLDY+pSQNwY23ixae8Q==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-AntiSpam-MessageData-ChunkCount:X-MS-Exchange-AntiSpam-MessageData-0:X-MS-Exchange-AntiSpam-MessageData-1;
 bh=7N8JCa2g2lf1em2rWSLx//NSbUzOTVVgGzz1o4lCCv4=;
 b=Q9rBdIxLyHFaQqIhVhQyRTXOnvxY4G198ReZcRUvWAlb0SyN2hH61gBaqYYlsifIbprefiqOr09eGXHgfyaVNuVBND1OVpa7jjwgUYmuWYea5e/MIfps7Wzk7FLQdN5tFKlUVFOpLII7jx6PEMj+JHCKJOEyn9dQ/ijskJZ8ftz5iC3XjTUj7gsA/eP0o+Q2Zi3u3DYBblztEyG65zV/Q6ElNcWGHc+uglR8rWRSVtrO0M0jKay+ZAgU1kll6G6+WUglBbDIXYXJPgcQDCoUj2xlU/Y7tM0vyAtdnOZudmeLqAIKLvs1ONllLwa0zfKYc49sBcA90ALX3w3IMB4Oyw==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=7N8JCa2g2lf1em2rWSLx//NSbUzOTVVgGzz1o4lCCv4=;
 b=Mz5+6dXhB8wsT2NAj8Qcnzta6O11ktDLwzgpYZrUustDnGZE7j1Bg6yRxw9w6cTnugXX4SP06k/JN19R5/SbjbFAfuCxh8U/hs5mytTDm76X5qnuCAyHiJqgiqr7RV5OUJfMAOuv219hcfBehU4pePrj92yGSewSTRW+573coVA=
Received: from SN6PR10MB3022.namprd10.prod.outlook.com (2603:10b6:805:d8::25)
 by CH0PR10MB4876.namprd10.prod.outlook.com (2603:10b6:610:c9::16) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.5588.15; Tue, 6 Sep
 2022 19:49:49 +0000
Received: from SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223]) by SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223%5]) with mapi id 15.20.5588.018; Tue, 6 Sep 2022
 19:49:49 +0000
From: Liam Howlett <liam.howlett@oracle.com>
To: "maple-tree@lists.infradead.org" <maple-tree@lists.infradead.org>,
        "linux-mm@kvack.org" <linux-mm@kvack.org>,
        "linux-kernel@vger.kernel.org"
	<linux-kernel@vger.kernel.org>,
        Andrew Morton <akpm@linux-foundation.org>
CC: Liam Howlett <liam.howlett@oracle.com>, Vlastimil Babka <vbabka@suse.cz>
Subject: [PATCH v14 58/70] mm/mprotect: use maple tree navigation instead of
 VMA linked list
Thread-Topic: [PATCH v14 58/70] mm/mprotect: use maple tree navigation instead
 of VMA linked list
Thread-Index: AQHYwim3KUNzm8XkwUqHosmeqJgucA==
Date: Tue, 6 Sep 2022 19:49:02 +0000
Message-ID: <20220906194824.2110408-59-Liam.Howlett@oracle.com>
References: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
In-Reply-To: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-mailer: git-send-email 2.35.1
x-ms-publictraffictype: Email
x-ms-office365-filtering-correlation-id: 922bec4a-fa81-4948-82b1-08da9040f51a
x-ms-traffictypediagnostic: CH0PR10MB4876:EE_
x-ms-exchange-senderadcheck: 1
x-ms-exchange-antispam-relay: 0
x-microsoft-antispam: BCL:0;
x-microsoft-antispam-message-info: 
 UgB/E+jTC/sfK7RZMWrwbn+IZRlQH+Cj06UUlexNKe0Y7VLMYcikkRmCArTvmooAncEM03DeAZgChx4v+qXqVVAPSIEgij5xVbl/jhAlF5poDHsfHHCHB7uQMNktb3u3u8OHUOnGv/ukxt4dL1BTd5S5A3M2kusd1JQPFU6+7bFKpULqo3ddDHwhIaHqZnh+kMKJ3HJUblPe1awiijbcJzPZvKbiQEnqkG90uP9i2nJZ26hqc7gKU1J+LfoqB/OJc617Nr6QcQbDmgg6yUWXyBMsqoDD++lb5yIO7RMHpRTy094pb2608YWXWxBEPDa5FNPfOXL7B251lLlie7xmPF1KVtfmmN6vym5UA2sA7meqeY0t5UL3oSBjDtBM9r6uMxcQvnJQGrmuuVsjnSn8OVPUS5DGo3+5uJ79etfHQxptxlcWoipLSM4ESHfj3hNXo2q/jzODUUfy05J0+rs5CysrkC+jfY7JwHsyA9YEjSQR9MwHhqrkv9zAmf0KueD5s2YGqopPNqdw/RTE7/lipOBWAeABDTGzmnS9YZbr4pUcj8Aiwm1iCc0JU5vp/eNjkAa2c0fUnsFq/1Cjqx10yhsN0i4tl/h2vsjtYMHjLVkhmqquP8+b2G3q3IQc0eCkF1vy0wy3wsEX5oKljdSRgqOtJkiNoh9FZDBKIJKvq6GrfjCh3xTyJrB43dUnNTF19bHbEyJe8XQN1gtyxmSgloVShYSoENHWSMay/k4Tlo89K5pc9DCYxwIDSxT9iKVm/7wQnV+PQ4/C6YF4YG22yA==
x-forefront-antispam-report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3022.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(13230016)(39860400002)(136003)(366004)(376002)(396003)(346002)(8936002)(6486002)(8676002)(4326008)(91956017)(122000001)(66946007)(66446008)(186003)(66476007)(76116006)(64756008)(66556008)(2616005)(38100700002)(38070700005)(478600001)(83380400001)(36756003)(1076003)(5660300002)(41300700001)(71200400001)(6666004)(44832011)(26005)(6506007)(54906003)(316002)(86362001)(2906002)(110136005)(6512007);DIR:OUT;SFP:1101;
x-ms-exchange-antispam-messagedata-chunkcount: 1
x-ms-exchange-antispam-messagedata-0: =?iso-8859-1?q?G2vISEqmgW05EnBIXyeW9S3?=
	=?iso-8859-1?q?w/xwUP89O2TMy1U/uY54Bv+LxESTypl6aZ1oYRabhwuh3EyR+fUW9Co/gKad?=
	=?iso-8859-1?q?0CcS4z96fY+mG0Fj0k0leijobXlKolIqYKJX9JKIBGH+/123Ce2wL/g+Oyr7?=
	=?iso-8859-1?q?UbDTSD7QZ/G2zcbOKvQWz2Z1yCpcYorJplMQVSSzzzinIlSxDcMo5P3MNdgl?=
	=?iso-8859-1?q?P/DYfrcSjekUDFKRfpHYi6hPvSMUgfIWI7DkXrPjme+SfE+ioAlfu034NObo?=
	=?iso-8859-1?q?IOPEXsfJC8Fsy/Y/uo32trmDYne6A+3fdxnfy/vn7W+buRt75iTZYBoJlvhG?=
	=?iso-8859-1?q?zJHt1Gu+nvcQKl62s1kMPJTka1oxAOonWlrqA3ywruQFV8DthDlguIzXDhQF?=
	=?iso-8859-1?q?K2mmuL2y/zKFXRqpwkXthyA1+9u4yTiLViJIJiPQYfJpAb97V0QJ2mvQI+RP?=
	=?iso-8859-1?q?e9YEjv+6L5H+wHkGHPPlnSkDGQQftSuhnwXbfzGFkLjgLsAaZnK3OWoWq8uH?=
	=?iso-8859-1?q?U/dKN9A66uv595He11xVHnhbeawVMw6V8XHWObctxODovNhqnXISpEb+r6Cp?=
	=?iso-8859-1?q?GF0qbt/4Fb3eyfVoOxiTQNKCvP3B9QPlM8ERVkV7Syh2cVAei9UPzOe42p/I?=
	=?iso-8859-1?q?hUdDZi0Jeaj+27YyEbW1JPrY6zR3rrMZmsRDJsmaYD32J/aLhu7Xn3M4BdFj?=
	=?iso-8859-1?q?8BkC/4wi6CYLMyyGO7YwODYySzogMl3B7R7Xi2aeZLD9zJuv3agA53EwTIel?=
	=?iso-8859-1?q?jD6K5DMjxc0dgBINOAOmlXsOGcJzLVHcnIqSSETSgceWzjRLbI/DuiSCyGP0?=
	=?iso-8859-1?q?YSNt+PLCVb9rQeM2XXsMjVRgYr3Q3UYCU54Z1WJ3FEyiQpCFl/Fq0GktBVsv?=
	=?iso-8859-1?q?cH6df/uONjUnt9LfW4LRI5EY9KBO8xk1FbDwUXEB9+UZGqCK/W00+FdkuM2k?=
	=?iso-8859-1?q?vQt4giqjqBrMAkJXSbDRu95wJEhxnH/oLqJ9vDx0cmskYTDnpqV8aE+UPU81?=
	=?iso-8859-1?q?OcOtmcrbnMBzc++Kwku6qjwISQphs4aTel+x/WQE5uVkNr5HgNrUf/qjaNAb?=
	=?iso-8859-1?q?Czqit6kXCpg2H/zfk/Zj68o2Cxsgogba8rVhdN32NEhNmtQzEBe0PRPuyNw1?=
	=?iso-8859-1?q?5miz0sXVPJClhBp1ZDiwadAihQLpt2u9EGgP3WLGfoZ0wtn4f0okXCboYw2c?=
	=?iso-8859-1?q?rrssolWpXgOF1S13N9fej4PyNwfKcBp1Y3NijJSN6yuQJm45L9GeJCeIsDon?=
	=?iso-8859-1?q?z6X67tG0PMjkkmiy54JtLLy7gA54QjM7dfzkJtJ9Mh1dMGanGpEIX9gvzPft?=
	=?iso-8859-1?q?cu1IatMqMapwPGoiiMt9vIQolvc5QKn4SIgjVZUvMZGgv8lffNhQOYuVCjQF?=
	=?iso-8859-1?q?KiY2prR2nktf15kLV9kEoggXePKMsMcoSOcaOBq49DV/BJOlNLiBJbhgctb0?=
	=?iso-8859-1?q?8xd/vcwLQMZ32D35PVTzDFa+/RtWpowc2T/pFo/gEqXog42aU/P7Up3FNBjr?=
	=?iso-8859-1?q?Ge+ooEiHdRntJAN/SthBF2jIpJH7XXBRlz/W/im2fUHcO0Gt1ljP1wwrUK8l?=
	=?iso-8859-1?q?E4YsvOsRqHV0X62l4PdhpFDjtYDtqstn92v+fAXvoNKaaVcirLcrku1QL5q3?=
	=?iso-8859-1?q?CrVnyJqYf+S7Y+hFW9361yVDis8aqDgC/AXTzbA=3D=3D?=
MIME-Version: 1.0
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3022.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 922bec4a-fa81-4948-82b1-08da9040f51a
X-MS-Exchange-CrossTenant-originalarrivaltime: 06 Sep 2022 19:49:02.9433
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-mailboxtype: HOSTED
X-MS-Exchange-CrossTenant-userprincipalname: 
 Ea1Ytkn9waMtN5XQGSgQ8Mk0y6s1XTti+ZBSpjIjVx4kEQ+5igUpsfjKQZvtLmmKyJR5I31lMb+arUPF4rGSVg==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: CH0PR10MB4876
X-Proofpoint-Virus-Version: vendor=baseguard
 engine=ICAP:2.0.205,Aquarius:18.0.895,Hydra:6.0.528,FMLib:17.11.122.1
 definitions=2022-09-06_09,2022-09-06_02,2022-06-22_01
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 spamscore=0
 malwarescore=0 mlxscore=0
 suspectscore=0 mlxlogscore=999 phishscore=0 adultscore=0 bulkscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2207270000
 definitions=main-2209060091
X-Proofpoint-ORIG-GUID: dd_WkEOeAXAkoyz_txnG8yyQTnrArhoC
X-Proofpoint-GUID: dd_WkEOeAXAkoyz_txnG8yyQTnrArhoC
ARC-Seal: i=2; s=arc-20220608; d=hostedemail.com; t=1662493793; a=rsa-sha256;
	cv=pass;
	b=siKVFv73dLw6Bn1urCUSY6kd2IiWIi7qOI1MthiKZzHrHUh0f/h5fjDrrlDon9s2C/443H
	oF1cL0lQJTAOytYHBbsg1+IE/MWRcbsa6ToVsNarTNVLs9ETEjWJmPbcRyGUV48OVBWinj
	26Wgfwq/kHkndxhfTyCH6MjB9dwCXPw=
ARC-Authentication-Results: i=2;
	imf14.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=hh3v+HEP;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=Mz5+6dXh;
	spf=pass (imf14.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	dmarc=pass (policy=none) header.from=oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1")
ARC-Message-Signature: i=2; a=rsa-sha256; c=relaxed/relaxed;
 d=hostedemail.com;
	s=arc-20220608; t=1662493793;
	h=from:from:sender:reply-to:subject:subject:date:date:
	 message-id:message-id:to:to:cc:cc:mime-version:mime-version:
	 content-type:content-type:
	 content-transfer-encoding:content-transfer-encoding:
	 in-reply-to:in-reply-to:references:references:dkim-signature;
	bh=7N8JCa2g2lf1em2rWSLx//NSbUzOTVVgGzz1o4lCCv4=;
	b=T2IJWAFX6kYRQ950ncBcPYRp9C6K0AAB9xNdGE0kQrKt8p6jizyz4hb59fZHc80a0oeGIx
	qjVLbioOPvWiPzfecOQ9AZgA10IglZJoMfClVAmBqHiJwDHagQbQ+NNI0YwH1aarl1iNxV
	q9d6ua6hLNYka3YwDvxC0gXLaV9oRR0=
Authentication-Results: imf14.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=hh3v+HEP;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=Mz5+6dXh;
	spf=pass (imf14.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	dmarc=pass (policy=none) header.from=oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1")
X-Rspam-User: 
X-Rspamd-Server: rspam06
X-Stat-Signature: wnd1eb61kdiwufekturc7trhacchyrun
X-Rspamd-Queue-Id: 70D30100069
X-HE-Tag: 1662493793-34687
X-Bogosity: Ham, tests=bogofilter, spamicity=0.000000, version=1.2.4
Sender: owner-linux-mm@kvack.org
Precedence: bulk
X-Loop: owner-majordomo@kvack.org
List-ID: <linux-mm.kvack.org>

From: "Liam R. Howlett" <Liam.Howlett@Oracle.com>

Switch to navigating the VMA list with the maple tree operators in
preparation for removing the linked list.

Signed-off-by: Liam R. Howlett <Liam.Howlett@Oracle.com>
Acked-by: Vlastimil Babka <vbabka@suse.cz>
---
 mm/mprotect.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/mm/mprotect.c b/mm/mprotect.c
index 1bdac9337fb9..3c79796be65e 100644
--- a/mm/mprotect.c
+++ b/mm/mprotect.c
@@ -675,6 +675,7 @@ static int do_mprotect_pkey(unsigned long start, size_t len,
 	const bool rier = (current->personality & READ_IMPLIES_EXEC) &&
 				(prot & PROT_READ);
 	struct mmu_gather tlb;
+	MA_STATE(mas, &current->mm->mm_mt, 0, 0);
 
 	start = untagged_addr(start);
 
@@ -706,7 +707,8 @@ static int do_mprotect_pkey(unsigned long start, size_t len,
 	if ((pkey != -1) && !mm_pkey_is_allocated(current->mm, pkey))
 		goto out;
 
-	vma = find_vma(current->mm, start);
+	mas_set(&mas, start);
+	vma = mas_find(&mas, ULONG_MAX);
 	error = -ENOMEM;
 	if (!vma)
 		goto out;
@@ -732,7 +734,7 @@ static int do_mprotect_pkey(unsigned long start, size_t len,
 	if (start > vma->vm_start)
 		prev = vma;
 	else
-		prev = vma->vm_prev;
+		prev = mas_prev(&mas, 0);
 
 	tlb_gather_mmu(&tlb, current->mm);
 	for (nstart = start ; ; ) {
@@ -795,7 +797,7 @@ static int do_mprotect_pkey(unsigned long start, size_t len,
 		if (nstart >= end)
 			break;
 
-		vma = prev->vm_next;
+		vma = find_vma(current->mm, prev->vm_end);
 		if (!vma || vma->vm_start != nstart) {
 			error = -ENOMEM;
 			break;

